orttraining/orttraining/python/training/orttrainer_options.py

```diff
@@ -92,6 +116,30 @@ class ORTTrainerOptions(object):
                         'enable_adasum' : {
                             'type' : 'boolean',
                             'default' : False
+                        },
+                        'sliced_schema' : {
+                            'type': 'dict',
```

Comment:
when `type: 'dict'`, a `schema` section must be specified to define the dict structure. `valuesrules` and `keysrules` cannot be defined outside `schema : {}`. Refer to `deepspeed_zero_optimization` for an example

Reply:
Sure. In my latest commit, we use
```python
            'sliced_schema' : {
                'type': 'dict',
                'required': False,
                'schema': {
                    # the keys in `sliced_schema` are strings.
                    'keysrules': {'type': 'string'},
                    # the values in `sliced_schema` are lists of integers.
                    'valuesrules': {
                        'type': 'list',
                        'schema': {'type': 'integer'}
                    }
                },
                'default_setter': lambda _: {}
            },
```
However, this doesn't fix the bug I found when assigning dictionary from Python to C++. Thus, I change `_wrap(...)` to
```python
    def _wrap(self, k, v):
        not_wrapped_fields = ['sliced_schema']
        if k in not_wrapped_fields or not isinstance(v, (tuple, list, set, frozenset, dict)):
            return v

        if isinstance(v, (tuple, list, set, frozenset)):
            return type(v)([self._wrap(k, v) for v in v])
        else:
            # This is a dict type.
            return _ORTTrainerOptionsInternal(self._main_class_name, v)

```

---
In reply to: [542786340](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r542786340) [](ancestors = 542786340)

Reply:
It seems the link is not valid.

Mind re-pasting a new link? Curious what the error is.

---
In reply to: [542957681](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r542957681) [](ancestors = 542957681,542786340)

Reply:
Type mis-match when assigning Python object to C++ object generated by pybind.

---
In reply to: [551520695](https://github.com/microsoft/onnxruntime/pull/5815#discussion_r551520695) [](ancestors = 551520695,542957681,542786340)